// ng-mat-search-bar/esm5/src/app/ng-mat-search-bar/util/abstract-value-accessor.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @abstract
 * @template T
 */
var /**
 * @abstract
 * @template T
 */
AbstractControlValueAccessor = /** @class */ (function () {
    function AbstractControlValueAccessor() {
        /**
         * Invoked when the model has been changed
         */
        this.onChange = (/**
         * @param {?} _
         * @return {?}
         */
        function (_) { });
        /**
         * Invoked when the model has been touched
         */
        this.onTouched = (/**
         * @return {?}
         */
        function () { });
    }
    /**
     * Method that is invoked on an update of a model.
     */
    /**
     * Method that is invoked on an update of a model.
     * @return {?}
     */
    AbstractControlValueAccessor.prototype.updateChanges = /**
     * Method that is invoked on an update of a model.
     * @return {?}
     */
    function () {
        this.onChange(this.value);
    };
    /**
     * Writes a new item to the element.
     * @param value the value
     */
    /**
     * Writes a new item to the element.
     * @param {?} value the value
     * @return {?}
     */
    AbstractControlValueAccessor.prototype.writeValue = /**
     * Writes a new item to the element.
     * @param {?} value the value
     * @return {?}
     */
    function (value) {
        this.value = value;
        this.updateChanges();
    };
    /**
     * Registers a callback function that should be called when the control's value changes in the UI.
     * @param fn
     */
    /**
     * Registers a callback function that should be called when the control's value changes in the UI.
     * @param {?} fn
     * @return {?}
     */
    AbstractControlValueAccessor.prototype.registerOnChange = /**
     * Registers a callback function that should be called when the control's value changes in the UI.
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.onChange = fn;
    };
    /**
     * Registers a callback function that should be called when the control receives a blur event.
     * @param fn
     */
    /**
     * Registers a callback function that should be called when the control receives a blur event.
     * @param {?} fn
     * @return {?}
     */
    AbstractControlValueAccessor.prototype.registerOnTouched = /**
     * Registers a callback function that should be called when the control receives a blur event.
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.onTouched = fn;
    };
    return AbstractControlValueAccessor;
}());
/**
 * @abstract
 * @template T
 */
export { AbstractControlValueAccessor };
if (false) {
    /** @type {?} */
    AbstractControlValueAccessor.prototype.value;
    /**
     * Invoked when the model has been changed
     * @type {?}
     */
    AbstractControlValueAccessor.prototype.onChange;
    /**
     * Invoked when the model has been touched
     * @type {?}
     */
    AbstractControlValueAccessor.prototype.onTouched;
}
