// ngx-particlesjs/esm5/lib/ngx-particles.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, ElementRef, HostListener, Input, ViewChild } from '@angular/core';
import { Particle } from './particle';
var NgxParticlesComponent = /** @class */ (function () {
    function NgxParticlesComponent() {
        var _this = this;
        this.particles = [];
        this.resizing = false;
        this.total = 100;
        this.stageChange = function () {
            var e_1, _a;
            _this.initStage();
            /** @type {?} */
            var range = [_this.width, _this.height];
            try {
                for (var _b = tslib_1.__values(_this.particles), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var x = _c.value;
                    x.onRange(range);
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_1) throw e_1.error; }
            }
            _this.resizing = false;
        };
        this.drawContext = function () {
            var e_2, _a;
            _this.ctx.clearRect(0, 0, _this.width, _this.height);
            try {
                for (var _b = tslib_1.__values(_this.particles), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var x = _c.value;
                    x.draw(_this.ctx);
                    x.update();
                }
            }
            catch (e_2_1) { e_2 = { error: e_2_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_2) throw e_2.error; }
            }
            _this.mainAnimationFrame = window.requestAnimationFrame(_this.drawContext);
            _this.resizing = false;
        };
    }
    /**
     * @return {?}
     */
    NgxParticlesComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        this.canvas = this.stageRef.nativeElement;
        this.ctx = this.canvas.getContext('2d', { alpha: false });
        this.initStage();
        this.initParticle();
        this.drawContext();
    };
    /**
     * @return {?}
     */
    NgxParticlesComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.ctx.clearRect(0, 0, this.width, this.height);
        window.cancelAnimationFrame(this.mainAnimationFrame);
        window.cancelAnimationFrame(this.changeAnimationFrame);
    };
    /**
     * @return {?}
     */
    NgxParticlesComponent.prototype.onResize = /**
     * @return {?}
     */
    function () {
        if (!this.resizing && (!this.default_height || !this.default_width)) {
            this.resizing = true;
            if (window.requestAnimationFrame) {
                window.requestAnimationFrame(this.stageChange);
            }
            else {
                setTimeout(this.stageChange, 66);
            }
        }
    };
    /**
     * @return {?}
     */
    NgxParticlesComponent.prototype.initStage = /**
     * @return {?}
     */
    function () {
        var _a = this.canvas.parentElement, clientWidth = _a.clientWidth, clientHeight = _a.clientHeight;
        if (this.default_width) {
            this.canvas.width = this.width = this.default_width;
        }
        else {
            this.canvas.width = this.width = clientWidth;
        }
        if (this.default_height) {
            this.canvas.height = this.height = this.default_height;
        }
        else {
            this.canvas.height = this.height = clientHeight;
        }
    };
    /**
     * @return {?}
     */
    NgxParticlesComponent.prototype.initParticle = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var range = [this.width, this.height];
        for (var i = 0; i < this.total; i++) {
            this.particles[i] = new Particle(range, (this.args && this.args.speed) ? this.args.speed : 'normal', (this.args && this.args.size) ? this.args.size : 1.5);
        }
    };
    NgxParticlesComponent.decorators = [
        { type: Component, args: [{
                    selector: 'ngx-particle',
                    template: "\n    <div id=\"stage\">\n      <canvas #stageRef></canvas>\n    </div>\n  ",
                    styles: ["\n    #stage {\n      height: 100%;\n      width: 100%;\n      will-change: all;\n    }\n  "]
                }] }
    ];
    NgxParticlesComponent.propDecorators = {
        stageRef: [{ type: ViewChild, args: ['stageRef',] }],
        total: [{ type: Input }],
        default_width: [{ type: Input }],
        default_height: [{ type: Input }],
        args: [{ type: Input }],
        onResize: [{ type: HostListener, args: ['window:resize',] }]
    };
    return NgxParticlesComponent;
}());
export { NgxParticlesComponent };
if (false) {
    /** @type {?} */
    NgxParticlesComponent.prototype.stageRef;
    /** @type {?} */
    NgxParticlesComponent.prototype.canvas;
    /** @type {?} */
    NgxParticlesComponent.prototype.ctx;
    /** @type {?} */
    NgxParticlesComponent.prototype.width;
    /** @type {?} */
    NgxParticlesComponent.prototype.height;
    /** @type {?} */
    NgxParticlesComponent.prototype.particles;
    /** @type {?} */
    NgxParticlesComponent.prototype.mainAnimationFrame;
    /** @type {?} */
    NgxParticlesComponent.prototype.changeAnimationFrame;
    /** @type {?} */
    NgxParticlesComponent.prototype.resizing;
    /** @type {?} */
    NgxParticlesComponent.prototype.total;
    /** @type {?} */
    NgxParticlesComponent.prototype.default_width;
    /** @type {?} */
    NgxParticlesComponent.prototype.default_height;
    /** @type {?} */
    NgxParticlesComponent.prototype.args;
    /** @type {?} */
    NgxParticlesComponent.prototype.stageChange;
    /** @type {?} */
    NgxParticlesComponent.prototype.drawContext;
}
