// ng-mat-search-bar/esm2015/src/app/ng-mat-search-bar/util/abstract-value-accessor.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @abstract
 * @template T
 */
export class AbstractControlValueAccessor {
    constructor() {
        /**
         * Invoked when the model has been changed
         */
        this.onChange = (/**
         * @param {?} _
         * @return {?}
         */
        (_) => { });
        /**
         * Invoked when the model has been touched
         */
        this.onTouched = (/**
         * @return {?}
         */
        () => { });
    }
    /**
     * Method that is invoked on an update of a model.
     * @return {?}
     */
    updateChanges() {
        this.onChange(this.value);
    }
    /**
     * Writes a new item to the element.
     * @param {?} value the value
     * @return {?}
     */
    writeValue(value) {
        this.value = value;
        this.updateChanges();
    }
    /**
     * Registers a callback function that should be called when the control's value changes in the UI.
     * @param {?} fn
     * @return {?}
     */
    registerOnChange(fn) {
        this.onChange = fn;
    }
    /**
     * Registers a callback function that should be called when the control receives a blur event.
     * @param {?} fn
     * @return {?}
     */
    registerOnTouched(fn) {
        this.onTouched = fn;
    }
}
if (false) {
    /** @type {?} */
    AbstractControlValueAccessor.prototype.value;
    /**
     * Invoked when the model has been changed
     * @type {?}
     */
    AbstractControlValueAccessor.prototype.onChange;
    /**
     * Invoked when the model has been touched
     * @type {?}
     */
    AbstractControlValueAccessor.prototype.onTouched;
}
