// ngx-particlesjs/esm2015/lib/ngx-particles.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, ElementRef, HostListener, Input, ViewChild } from '@angular/core';
import { Particle } from './particle';
export class NgxParticlesComponent {
    constructor() {
        this.particles = [];
        this.resizing = false;
        this.total = 100;
        this.stageChange = () => {
            this.initStage();
            /** @type {?} */
            const range = [this.width, this.height];
            for (const x of this.particles) {
                x.onRange(range);
            }
            this.resizing = false;
        };
        this.drawContext = () => {
            this.ctx.clearRect(0, 0, this.width, this.height);
            for (const x of this.particles) {
                x.draw(this.ctx);
                x.update();
            }
            this.mainAnimationFrame = window.requestAnimationFrame(this.drawContext);
            this.resizing = false;
        };
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        this.canvas = this.stageRef.nativeElement;
        this.ctx = this.canvas.getContext('2d', { alpha: false });
        this.initStage();
        this.initParticle();
        this.drawContext();
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.ctx.clearRect(0, 0, this.width, this.height);
        window.cancelAnimationFrame(this.mainAnimationFrame);
        window.cancelAnimationFrame(this.changeAnimationFrame);
    }
    /**
     * @return {?}
     */
    onResize() {
        if (!this.resizing && (!this.default_height || !this.default_width)) {
            this.resizing = true;
            if (window.requestAnimationFrame) {
                window.requestAnimationFrame(this.stageChange);
            }
            else {
                setTimeout(this.stageChange, 66);
            }
        }
    }
    /**
     * @return {?}
     */
    initStage() {
        const { clientWidth, clientHeight } = this.canvas.parentElement;
        if (this.default_width) {
            this.canvas.width = this.width = this.default_width;
        }
        else {
            this.canvas.width = this.width = clientWidth;
        }
        if (this.default_height) {
            this.canvas.height = this.height = this.default_height;
        }
        else {
            this.canvas.height = this.height = clientHeight;
        }
    }
    /**
     * @return {?}
     */
    initParticle() {
        /** @type {?} */
        const range = [this.width, this.height];
        for (let i = 0; i < this.total; i++) {
            this.particles[i] = new Particle(range, (this.args && this.args.speed) ? this.args.speed : 'normal', (this.args && this.args.size) ? this.args.size : 1.5);
        }
    }
}
NgxParticlesComponent.decorators = [
    { type: Component, args: [{
                selector: 'ngx-particle',
                template: `
    <div id="stage">
      <canvas #stageRef></canvas>
    </div>
  `,
                styles: [`
    #stage {
      height: 100%;
      width: 100%;
      will-change: all;
    }
  `]
            }] }
];
NgxParticlesComponent.propDecorators = {
    stageRef: [{ type: ViewChild, args: ['stageRef',] }],
    total: [{ type: Input }],
    default_width: [{ type: Input }],
    default_height: [{ type: Input }],
    args: [{ type: Input }],
    onResize: [{ type: HostListener, args: ['window:resize',] }]
};
if (false) {
    /** @type {?} */
    NgxParticlesComponent.prototype.stageRef;
    /** @type {?} */
    NgxParticlesComponent.prototype.canvas;
    /** @type {?} */
    NgxParticlesComponent.prototype.ctx;
    /** @type {?} */
    NgxParticlesComponent.prototype.width;
    /** @type {?} */
    NgxParticlesComponent.prototype.height;
    /** @type {?} */
    NgxParticlesComponent.prototype.particles;
    /** @type {?} */
    NgxParticlesComponent.prototype.mainAnimationFrame;
    /** @type {?} */
    NgxParticlesComponent.prototype.changeAnimationFrame;
    /** @type {?} */
    NgxParticlesComponent.prototype.resizing;
    /** @type {?} */
    NgxParticlesComponent.prototype.total;
    /** @type {?} */
    NgxParticlesComponent.prototype.default_width;
    /** @type {?} */
    NgxParticlesComponent.prototype.default_height;
    /** @type {?} */
    NgxParticlesComponent.prototype.args;
    /** @type {?} */
    NgxParticlesComponent.prototype.stageChange;
    /** @type {?} */
    NgxParticlesComponent.prototype.drawContext;
}
