// ngx-particlesjs/esm2015/lib/particle.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { saveDecimal } from './common';
export class Particle {
    /**
     * @param {?} range
     * @param {?=} speed
     * @param {?=} size
     */
    constructor(range, speed = 'normal', size = 1.5) {
        this.max_width = range[0];
        this.max_height = range[1];
        this.initPosition();
        this.initSpeed(speed);
        this.initSize(size);
        this.initColor();
    }
    /**
     * @param {?} range
     * @return {?}
     */
    onRange(range) {
        this.max_width = range[0];
        this.max_height = range[1];
    }
    /**
     * @return {?}
     */
    initPosition() {
        this.x = saveDecimal(Math.random() * this.max_width);
        this.y = saveDecimal(Math.random() * this.max_height);
    }
    /**
     * @param {?} speed
     * @return {?}
     */
    initSpeed(speed) {
        /** @type {?} */
        let vi = [];
        if (typeof speed === 'string') {
            switch (speed) {
                case 'fast':
                    vi = [1, 1];
                    break;
                case 'normal':
                    vi = [2, 2];
                    break;
                case 'slow':
                    vi = [5, 5];
                    break;
                default:
                    vi = [2, 2];
            }
        }
        else {
            vi = speed;
        }
        this.vx = saveDecimal(Math.random() / vi[0]);
        this.vy = saveDecimal(Math.random() / vi[1]);
    }
    /**
     * @param {?} size
     * @return {?}
     */
    initSize(size) {
        this.size = (1 + Math.random()) * size;
    }
    /**
     * @return {?}
     */
    initColor() {
        this.color = `rgb(${Math.floor(Math.random() * 256)},${Math.floor(Math.random() * 256)},${Math.floor(Math.random() * 256)})`;
    }
    /**
     * @param {?} ctx
     * @return {?}
     */
    draw(ctx) {
        ctx.beginPath();
        ctx.arc(this.x, this.y, this.size, 0, 2 * Math.PI);
        ctx.fillStyle = this.color;
        ctx.strokeStyle = this.color;
        ctx.closePath();
        ctx.fill();
        ctx.stroke();
    }
    /**
     * @return {?}
     */
    update() {
        this.edge();
        this.x = this.x + this.vx;
        this.y = this.y + this.vy;
    }
    /**
     * @return {?}
     */
    edge() {
        if (this.x + this.vx > this.max_width || this.x + this.vx < 0) {
            this.vx = -this.vx;
        }
        if (this.y + this.vy > this.max_height || this.y + this.vy < 0) {
            this.vy = -this.vy;
        }
    }
}
if (false) {
    /** @type {?} */
    Particle.prototype.max_width;
    /** @type {?} */
    Particle.prototype.max_height;
    /** @type {?} */
    Particle.prototype.x;
    /** @type {?} */
    Particle.prototype.y;
    /** @type {?} */
    Particle.prototype.vx;
    /** @type {?} */
    Particle.prototype.vy;
    /** @type {?} */
    Particle.prototype.size;
    /** @type {?} */
    Particle.prototype.color;
}
